checker/src/main/java/org/checkerframework/checker/resourceleak/MustCallInference.java

```diff
@@ -1,176 +1,845 @@
+   * @param cfg the control flow graph of the method to check
+   * @param mcca the MustCallConsistencyAnalyzer
+   */
+  /*package-private*/ static void runMustCallInference(
+      ResourceLeakAnnotatedTypeFactory resourceLeakAtf,
+      ControlFlowGraph cfg,
+      MustCallConsistencyAnalyzer mcca) {
+    MustCallInference mustCallInferenceLogic = new MustCallInference(resourceLeakAtf, cfg, mcca);
+    mustCallInferenceLogic.runInference();
+  }
+
+  /**
+   * Runs the inference algorithm on the current method (the {@link #cfg} field). It updates the
+   * {@link #disposedFields} set or adds @Owning to the formal parameter if it discovers their
    * must-call obligations were satisfied along one of the checked paths.
+   *
+   * <p>Operationally, it checks method invocations for fields and parameters with
+   * non-empty @MustCall obligations along all paths to the regular exit point.
    */
-  /*package-private*/ void runInference() {
-    Set<Block> visited = new HashSet<>();
-    Deque<Block> worklist = new ArrayDeque<>();
-    Block entry = this.cfg.getEntryBlock();
+  private void runInference() {
 
+    Set<BlockWithObligations> visited = new HashSet<>();
+    Deque<BlockWithObligations> worklist = new ArrayDeque<>();
+
+    BlockWithObligations entry =
+        new BlockWithObligations(cfg.getEntryBlock(), getNonEmptyMCParams(cfg));
     worklist.add(entry);
     visited.add(entry);
 
     while (!worklist.isEmpty()) {
-      Block current = worklist.remove();
+      BlockWithObligations current = worklist.remove();
+
+      Set<Obligation> obligations = new LinkedHashSet<>(current.obligations);
 
-      for (Node node : current.getNodes()) {
+      for (Node node : current.block.getNodes()) {
+        // The obligation set calculated for RLC differs from the Inference process. In the
+        // Inference process, it exclusively tracks parameters with non-empty must-call types,
+        // whether they have the @Owning annotation or not. However, there are some shared
+        // computations, such as 'updateObligationsWithInvocationResult,' which is used during
+        // inference and could potentially affect the RLC result if it were called before the
+        // checking phase. However, calling 'updateObligationsWithInvocationResult()' will not have
+        // any side effects on the outcome of the Resource Leak Checker. This is because the
+        // inference occurs within the postAnalyze method of the ResourceLeakAnnotatedTypeFactory,
+        // once the consistency analyzer has completed its process
         if (node instanceof MethodInvocationNode) {
-          checkForMustCallInvocationOnField((MethodInvocationNode) node);
+          mcca.updateObligationsWithInvocationResult(obligations, node);
+          computeOwningFromInvocation(obligations, (MethodInvocationNode) node);
+        } else if (node instanceof ObjectCreationNode) {
+          mcca.updateObligationsWithInvocationResult(obligations, node);
```

Comment:
Should we be running logic like `inferOwningParamsViaOwnershipTransfer()` for an `ObjectCreationNode`?  It seems like right now we will miss inferring an annotation in a case like the following:
```java
class Foo {
  Foo(@Owning Socket s) { ... }
}
class Bar {
  static void baz(Socket s) {
    Foo f = new Foo(s);
  }
}
```
It seems like we should be able to infer `@Owning` for the parameter `s` of `baz` but right now we won't?

In fact, an `ObjectCreationNode` represents a type of method invocation, so in principle other inference actions inside `computeOwningFromInvocation()` might apply here (though they may be unlikely to apply in practice).